lib/awful/wibar.lua

```diff
@@ -228,22 +349,99 @@ local function set_position(wb, position, skip_reattach)
     wb:emit_signal("property::position", position)
 end
 
-local function get_stretch(w)
+function awfulwibar.get_stretch(w)
     return w._stretch
 end
 
-local function set_stretch(w, value)
+function awfulwibar.set_stretch(w, value)
     w._stretch = value
 
     attach(w, w.position)
 
     w:emit_signal("property::stretch", value)
 end
 
+
+function awfulwibar.get_update_workarea(w)
+    return w._private.update_workarea
+end
+
+function awfulwibar.set_update_workarea(w, value)
+    w._private.update_workarea = value
+
+    attach(w, w.position)
+
+    w:emit_signal("property::update_workarea", value)
+end
+
+
+function awfulwibar.set_margins(w, value)
+    if type(value) == "number" then
+        value = {
+            top = value,
+            bottom = value,
+            right = value,
+            left = value,
+        }
+    end
+
+   value = value or {
+        left   = 0,
+        right  = 0,
+        top    = 0,
+        bottom = 0
+    }
+
+    w._private.margins = value
+
+    attach(w, w.position)
+
+    w:emit_signal("property::margins", value)
+end
+
+-- Allow each margins to be set individually.
+local function meta_margins(self)
+    return setmetatable({}, {
+        __index = self._private.margins,
+        __newindex = function(_, k, v)
+            self._private.margins[k] = v
+            awfulwibar.set_margins(self, self._private.margins)
+        end
+    })
+end
+
+function awfulwibar.get_margins(self)
+    return self._private.meta_margins
+end
```

Comment:
I am conflicted this this. One way, this brings us closer to QML level of syntax sugar. On the other side, this leaves Lua 5.1 behind. Those meta magic have a big drawback. `pairs()` and `ipairs()` don't work on 5.1 and I don't know how we could fix it beyond monkey-patching `pairs()` and `ipairs()` in both the C and example generation. That might sound like a pretty minor problem, but it actually is worst because LGI and CAPI could not use the `args` if everything was built using `__index` layers.

I added it here because the margins need to be easily enabled/disabled using keybindings and that helps a lot:

```
awful.key({ modkey }, "d", function() s.mywibox.margins.bottom = 0 end),
```

But I am uncertain if this should be globally implemented.

Reply:
(sorry for the long delay, as usual)

> I would, instead, propose an API similar to CSS

I would prefer to avoid doing this. Having overlapping properties is quite confusing on its own. I kind of like the "property namespace" the table provides. Plus, the margin concept is used in many places already and use a table. Another concern I have is that it would make the documentation much larger and intimidating. Note that there is precedents to adding properties which are also in table. Some objects have both `:geometry()` and `x`/`y`/`width`/`height`. However that was a "consistency" hack to bring the various wibox/client like API closer to each other (just like I am doing with the wallpaper and dynamic client layout PRs).

I would prefer one of those solution:

 * Remove the meta margin code
 * Keep the meta margin code as-is
 * Refactor to make the meta margin code part of an `gears.object` and begin to integrate it in more places as @Aire-One suggests

I don't really mind which way this goes. There is pros and cons in each. @Aire-One @sclu1034 what's your vote?

> inconsistent with what the property::margins 

Along with the `pairs()` problem, there is some corner case where it will be inconvenient to be perfectly consistent.

Reply:
I honestly don't see the first issue raised by @sclu1034 as an unwanted behavior. Yes, if you use the same table reference for multiple objects properties, when you alter this table, all objects are affected. To prevent the side effects from the example, the easiest trick would be to clone the margins tables before providing it to the widget.

Hm... I think we can do the clone in the API, and provide it to the user? `__index = function () return gears.clone(self._private.margins) end`.